### PR TITLE
Specs for Offspring controller and model. Includes minor changes in other files

### PR DESCRIPTION
**Pull request by alexvkcr**
This PR includes the necessary specs to ensure the proper functionality of the Offspring controller and the Offspring model.

There are changes in other files that are necessary to pass the tests. 

This PR is a WIP and comments go back and forth, so they are important to find out what is being developed at the moment, 

@sergio-ocon @katafira @delaluzparra

### CODE REVIEW

**Comment by sergio-ocon** [comment]
Please review and then assign to me when reviewed @alexvkcr @delaluzparra

**Comment by sergio-ocon** [comment]
Hi @delaluzparra as @katafira is not here, can you review?

File: app/models/shift.rb [hunk 1]
```diff
@@ -23,7 +23,7 @@ def sites_available
   private
 
   def always_same_week
-    self.day = day_of_week
+    # self.day = day_of_week
```
Comment (alexvkcr) [inline app/models/shift.rb]: This line makes seeding impossible, and is apparently useless

**Comment by alexvkcr** [comment]
Assigned to me @alexvkcr to fix some mistakes.

Afterwards, it will be assigned to somebody else for review

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -15,4 +15,12 @@ def configure_permitted_parameters
     devise_parameter_sanitizer.for(:sign_up) << :last_name
     devise_parameter_sanitizer.for(:sign_up) << :phone
   end
+
+  def authenticate_user!
```
Comment (alexvkcr) [inline app/controllers/application_controller.rb]: This method is a temporary solution to pass the tests, pending to fix devise to redirect

**Comment by alexvkcr** [comment]
Tests have been fixed to request a redirection when not logged in.

**Comment by delaluzparra** [comment]
Please fill up the description about what this PR solves in order to review it clearly.

File: app/models/shift.rb (hunk 1 above)
Comment (sergio-ocon) [inline app/models/shift.rb]: Why does it make seeding impossible? We are seeding, and if you don't do this there will be an error if you don't set up the day

**Comment by sergio-ocon** [comment]
Purpose or Intent
-----------------
> Describe the rationale and use case for this pull request.  Provide any background, examples, and images that provide further information to accurately describe what it is that you are adding to the repo.  Add subsections as necessary to organize and feel free to link and reference other PRs as necessary, but also include them in the links section below as a quick reference.
> 
> Guidelines:
>   * Keep Pull Request titles short and to the point, ideally under 72 characters
>   * Provide as much context/info in the description as necessary to get the reviewer up to the same domain knowledge level as yourself
>   * Keep code changes as short and implementing a single feature/bug fix/refactoring when possible


Links
-----
> * http://documentation.for/library/that/I/am/adding
> * [PR relevant issue or pull_request](#123)


Steps for Testing/QA
--------------------
> [Optional] If there are any manual steps that you would like the reviewer(s) to take to verify your changes, please describe in detail the steps to reproduce the features added by the pull request, or the bug before and after the change.

**Comment by sergio-ocon** [comment]
El template para la descripción del Pull request

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (sergio-ocon) [inline app/controllers/application_controller.rb]: This can't be done, as you are overwriting devise authenticate_user!

You would need to create a new authentication funtion that calls authenticate user or authenticate admin and that takes into account more things
https://github.com/plataformatec/devise/wiki/How-To:-Define-resource-actions-that-require-authentication-using-routes.rb

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -15,4 +15,12 @@ def configure_permitted_parameters
...
+    begin
+      super
```
Comment (sergio-ocon) [inline app/controllers/application_controller.rb]: This is hiding the error.
Even if you do this because you don't find any other method you should document it throughly

File: app/controllers/offsprings_controller.rb [hunk 4]
```diff
@@ -11,7 +11,8 @@ def new
   def create
     @offspring = current_user.offsprings.build(offsprings_params)
     unless @offspring.primary_first?
-      redirect_to static_pages_intructions_path && return
+      redirect_to static_pages_intructions_path
```
Comment (sergio-ocon) [inline app/controllers/offsprings_controller.rb]: My code is using AND and it works.... why this is showing &&?
Also, it would be great to add a message so the user knows what is going on

File: app/models/shift.rb [hunk 5]
```diff
@@ -23,7 +23,7 @@ def sites_available
   private
 
   def always_same_week
```
Comment (sergio-ocon) [inline app/models/shift.rb]: always same week was used to make sure that the day of week was Monday, it was used to assign it to a year that started on Monday (Jan 1st was Monday). Is that needed any more? Now we don't use datetime I believe

File: spec/controllers/offsprings_controller_spec.rb [hunk 6]
```diff
@@ -2,25 +2,61 @@
...
-        expect(assigns).to eq(nil) #it does not go through the controller
-        expect(response).to redirect_to(new_user_session_path)
-      end
+
+    it "does not allow access new" do
+      get :new, {}
+      expect(response).to redirect_to('home')
+    end
+
+    let(:user) { FactoryGirl.create(:user) }
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: This is a lazy creation, you won't execute this code if it is not necessary, and that's why it is always at the beginning of the context block...

File: spec/controllers/offsprings_controller_spec.rb [hunk 7]
```diff
@@ -2,25 +2,61 @@
...
+          post :create, offspring: {first_name: "pepe", last_name: "kata", grade: :primary_first}
+        }.to change(user.offsprings, :count).by(1)
       end
+      it "does not allow any other and redirects" do
+        Offspring.grades.keys.each do |i|
+          if i != 'primary_first'
+            expect{
+              post :create, offspring: {first_name: "pepe", last_name: "kata", grade: i}
+            }.to change(user.offsprings, :count).by(0)
+            expect(response).to redirect_to(static_pages_intructions_path)
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: Please test also that a warning / error message is shown

File: spec/models/offspring_spec.rb [hunk 8]
```diff
@@ -7,43 +7,40 @@
   end
 
   it "is invalid without a name" do
-    child = FactoryGirl.build(:offspring, first_name:nil)
+    child = FactoryGirl.build(:offspring, first_name: nil)
```
Comment (sergio-ocon) [inline spec/models/offspring_spec.rb]: if you define a user with let, you can just change first_name here to nil, you don't need to create a new one

File: spec/models/offspring_spec.rb [hunk 9]
```diff
@@ -7,43 +7,40 @@
...
 
   it "is invalid without a surname" do
-    child = FactoryGirl.build(:offspring, last_name:nil)
+    child = FactoryGirl.build(:offspring, last_name: nil)
     expect(child).not_to be_valid
     child.valid?
     expect(child.errors[:last_name]).to include("no puede estar vacío")
   end
 
-  it "should have the same surname than the brothers" do
```
Comment (sergio-ocon) [inline spec/models/offspring_spec.rb]: Why did you delete this check? It is needed to make sure that all the brothers have the same surname

File: app/models/shift.rb (hunk 1 above)
Comment (alexvkcr) [inline app/models/shift.rb]: You ara giveing value to 'day' which  doesn't appear in the validations of the model.

Yesterday, when @delaluzparra tried to seed the database, there was an error in that line that dissapeared after commenting it.

It seems we're not going to use 'day' anymore, but before deleting any line I prefer to post a question

File: app/models/shift.rb (hunk 5 above)
Comment (alexvkcr) [inline app/models/shift.rb]: It seems we don't, We'll delete it

File: app/controllers/offsprings_controller.rb (hunk 4 above)
Comment (alexvkcr) [inline app/controllers/offsprings_controller.rb]: On https://github.com/sergio-ocon/lastablas the same file contains &&, not AND. You probably changed it locally and forgot it.

I'll add the message for the users!

File: spec/models/offspring_spec.rb (hunk 9 above)
Comment (alexvkcr) [inline spec/models/offspring_spec.rb]: Coming back up!

**Comment by sergio-ocon** [comment]
https://github.com/sergio-ocon/lastablas/tree/delete_offspring

**Comment by sergio-ocon** [comment]
You need to rebase....

File: app/models/shift.rb [hunk 10]
```diff
@@ -10,7 +10,6 @@ class Shift < ActiveRecord::Base
   validates_each :prebooked do |shift, _room|
     shift.errors.add(:shift, "no puede ser mayor que la capacidad de la sala") unless shift.sites_available?
   end
```
Comment (sergio-ocon) [inline app/models/shift.rb]: This has been erased already in the latest version

**Comment by alexvkcr** [comment]
We have been severely stuck at creating the spec 
**"should have the same surname than the brothers"**

The problem we had is that in any way we tried to access from an offspring the list of its brothers, we would always get an empty ActiveRecord.

We were executing this code in the spec
```
let (:parent) { FactoryGirl.build(:user, :last_name => 'Salado') }
let (:child) { FactoryGirl.build(:offspring, :user => parent) }
```

and we expected Factory Girl to include it in the list, and write in the user_id (field from  the id of parent.

Debugging deeply, we saw that Factory Girl ignores that association and creates both records independently.

Now that we know this, we have solved it and we have this spec ready to review. 

Assigning to @katafira

**Comment by alexvkcr** [comment]
Ready o review, waiting for Travis and @katafira

**Comment by katafira** [comment]
<h2>Looks good to me.</h2>

**Comment by sergio-ocon** [comment]
This is not WIP.

You need to rebase.

You need to address all remaining comments

**Comment by sergio-ocon** [comment]
FactoryGirl with let does not create it until it is used, and you are building it, so there is no associted user created.

1. You can create the user before and then associate the offspring to it.
2. You can create the offspring (not build it), so the user associated is created (because is needed by the factory)
3. You can save the offspring child.save before creating the brother...

I believe the third one is the best one, as it does not depend on implementation (the offspring needs to be associated to the user but you don't test the connection here), but you can choose anything

File: spec/controllers/offsprings_controller_spec.rb [hunk 11]
```diff
@@ -2,38 +2,75 @@
 # include Devise::TestHelpers # this should work without this
 
 RSpec.describe OffspringsController, type: :controller do
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: With Rails 5, controller spec are deprecated and substituted by request specs.... 
Also, there should be a message and you should test for it
flash[:error]= "El usuario tiene que estar autenticado" o algo parecido al fallar la autenticación

File: spec/controllers/offsprings_controller_spec.rb [hunk 12]
```diff
@@ -2,38 +2,75 @@
...
+        delete :destroy, id: off.id
+      end.to change(user.offsprings, :count).by(0)
+      expect(response).to redirect_to('home')
     end
 
-    context "When authenticated" do
-      pending "shows index of offspring" do
-        get :index
-        expect(response).to have_http_status(:success)
+    pending "does not allow to see the index of offsprings"
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: I would suggest you test this together:
actions.each do |action|
   expect.... to redirect to....

end

File: spec/controllers/offsprings_controller_spec.rb (hunk 11 above)
Comment (alexvkcr) [inline spec/controllers/offsprings_controller_spec.rb]: What do you mean by "are deprecated and substituted by request specs"? Should we change to Rails 5 style or you are just letting us know?

File: spec/models/offspring_spec.rb [hunk 13]
```diff
@@ -1,43 +1,50 @@
 require 'rails_helper'
 
 RSpec.describe Offspring, type: :model do
+  let(:child) { FactoryGirl.build(:offspring) }
```
Comment (sergio-ocon) [inline spec/models/offspring_spec.rb]: This has been refactored to make it easier to read

File: spec/controllers/offsprings_controller_spec.rb (hunk 11 above)
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: Both.

They will be available in ManageIQ, but you need to add a gem. 
This is ok, but it would be better to write request specs if you can find the documentation

Comment (alexvkcr) [inline spec/controllers/offsprings_controller_spec.rb]: I think there is already enough ongoing work, but it should be definitively included in TODO file something like "Changing specs to Rails 5 style. Replacing controller specs for request specs"

**Comment by alexvkcr** [comment]
There are two errors in rubocop that don't belong to this PR, what to do?

File: spec/models/offspring_spec.rb [hunk 14]
```diff
@@ -27,20 +27,23 @@
     expect(brother.errors[:last_name]).to include("tiene que coincidir con el de sus hermanos")
   end
 
-  it "is invalid wihout a grade" do
+  it "is invalid wihout an course" do
```
Comment (sergio-ocon) [inline spec/models/offspring_spec.rb]: a course?

File: spec/models/offspring_spec.rb [hunk 15]
```diff
@@ -27,20 +27,23 @@
...
     child.grade = nil
+    expect(child).not_to be_valid
     child.valid?
     expect(child.errors[:grade]).to include("no puede estar vacío")
   end
 
-  it "is invalid with grades that don't belong to the grade list" do
-    expect { child.grade = 100 }.to raise_error(ArgumentError).with_message(/is not a valid grade/)
+  it 'is in any grade' do
```
Comment (sergio-ocon) [inline spec/models/offspring_spec.rb]: Don't delete the error when the grade is not in the enum

File: spec/models/offspring_spec.rb [hunk 16]
```diff
@@ -27,20 +27,23 @@
...
+      child.grade = i
+      expect(child).to be_valid
+    end
   end
 
   it "is invalid without a user associated" do
     child.user = nil
-    child.valid?
-    expect(child.errors[:user]).to include("no puede faltar")
+    expect(child).not_to be_valid
```
Comment (sergio-ocon) [inline spec/models/offspring_spec.rb]: You should check for the error as I do, not the generic invalidity (that is part of the other test).
With only a check this will be faster and you are checking the same

File: spec/controllers/offsprings_controller_spec.rb (hunk 12 above)
Comment (alexvkcr) [inline spec/controllers/offsprings_controller_spec.rb]: I was writing the spec to test that all actions redirect correctly and I reached the conclusion that since each action requires a different verb, that loop would be fraught with 'ifs' for each actions. And so, I think it's better to check each redirect afeter checking its functionality.

Besides, in this way we will call only once every action, and with the suggested loop we'd call them twice.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (sergio-ocon) [inline app/controllers/application_controller.rb]: Please delete this. This PR IS NOT MERGEABLE WITH THIS.
Review the link and act as it says.

THERE WON'T BE ANY MERGE WITH THIS CODE IN IT.

File: app/controllers/offsprings_controller.rb [hunk 17]
```diff
@@ -25,7 +25,7 @@ def create
   def destroy
     offspring = Offspring.find_by_id(params[:id])
     offspring.destroy
-    flash[:sucess] = "Niño borrado"
+    flash[:success] = "Niño borrado"
```
Comment (sergio-ocon) [inline app/controllers/offsprings_controller.rb]: :+1:

File: spec/controllers/offsprings_controller_spec.rb (hunk 12 above)
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: Makes sense

File: spec/models/offspring_spec.rb [hunk 18]
```diff
@@ -2,9 +2,7 @@
 
 RSpec.describe Offspring, type: :model do
   let(:child) { FactoryGirl.build(:offspring) }
-
   it "has a valid factory" do
-    child.valid?
     expect(child).to be_valid
```
Comment (sergio-ocon) [inline spec/models/offspring_spec.rb]: if you don't test validity, I believe the test will always pass because tehre are nor errors in child.errors

**Comment by sergio-ocon** [comment]
Great improvement but you still need to review the comments

**Comment by sergio-ocon** [comment]
You can test with build_stubbed instead of build or create if you need an ID but you don't need to hit the database

File: spec/controllers/offsprings_controller_spec.rb [hunk 19]
```diff
@@ -2,38 +2,78 @@
...
+    it "does not allow access new" do
+      get :new, {}
+      expect(response).to redirect_to('home')
+    end
+
+    let(:user) { FactoryGirl.create(:user) }
+    let(:off) { FactoryGirl.create(:offspring, user: user) }
+    it "does not allow to destroy" do
+      user.save
+      off.save
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: This is not necessary as you are using create

**Comment by sergio-ocon** [comment]
Please review your code. It has so many create! that is really slow
Use build_dubbed so it does not hit the database unless is needed.
Don't put so many create when are not needed.

Make it faster!

File: spec/controllers/offsprings_controller_spec.rb [hunk 20]
```diff
@@ -2,38 +2,81 @@
...
-      pending "visitor can't see the index of offsprings"
-      pending "visitor can't see offpsring@show"
-      pending "visitor can't see other users"
+    it "does not allow access new" do
+      get :new, {}
+      expect(response).to redirect_to('/users/sign_in')
+    end
+
+    it "does not allow to destroy" do
+      user = FactoryGirl.build(:user)
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: This is already defined in the let(:user) at the beginning

File: spec/controllers/offsprings_controller_spec.rb [hunk 21]
```diff
@@ -2,38 +2,81 @@
...
+      user.save
+      off = FactoryGirl.build(:offspring, user: user)
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: build_stubbed

File: spec/controllers/offsprings_controller_spec.rb [hunk 22]
```diff
@@ -2,38 +2,81 @@
...
-    context "When admin authenticated" do
-      pending "can see offspring data"
-      pending "can delete any offspring"
-      pending "can create offpsring to any user"
+    it "#new shows view for new offspring" do
+      get :new, {}
+      expect(response).to render_template(:new)
     end
+    describe "#destroy" do
+      let(:off) { FactoryGirl.create(:offspring, user: user) } # Needed to be persistant
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: If you put the let at the beginning and here you save you get the same and you don't create two elements

File: spec/spec_helper.rb [hunk 23]
```diff
@@ -20,7 +20,6 @@
   # rspec-expectations config goes here. You can use an alternate
   # assertion/expectation library such as wrong or the stdlib/minitest
   # assertions if you prefer.
-  config.include Devise::TestHelpers, type: :controller
   config.include Warden::Test::Helpers, type: :request
```
Comment (sergio-ocon) [inline spec/spec_helper.rb]: I believe this is not necessary

File: spec/controllers/offsprings_controller_spec.rb [hunk 24]
```diff
@@ -2,38 +2,77 @@
...
+    it "does not allow access new" do
+      get :new, {}
+      expect(response).to redirect_to('/users/sign_in')
+    end
+
+    it "does not allow to destroy" do
+      expect do
+        delete :destroy, id: off.id
+      end.to change(user.offsprings, :count).by(0)
+      expect(response).to redirect_to('/users/sign_in')
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: Use it properly:
new_user_session_path
new_user_session_route

File: spec/controllers/offsprings_controller_spec.rb [hunk 25]
```diff
@@ -2,38 +2,77 @@
...
+    end
+
+    pending "does not allow to see the index of offsprings"
```
Comment (sergio-ocon) [inline spec/controllers/offsprings_controller_spec.rb]: Execute rake routes to see routes
It should be the same with a different URL, try to access them and be redirected.
